usaspending_api/disaster/v2/views/elasticsearch_base.py

```diff
@@ -0,0 +1,136 @@
+    sum_column_mapping: Dict[str, str]  # Overwritten by a pagination mixin
+
+    @cache_response()
+    def post(self, request: Request) -> Response:
+        # Handle "query" filter outside of QueryWithFilters
+        query = self.filters.pop("query", None)
+
+        # If for a loans endpoint then filter on Loans
+        filters = self.filters
+        if self.is_loans:
+            filters.update({"award_type_codes": list(loan_type_mapping)})
+        self.filter_query = QueryWithFilters.generate_awards_elasticsearch_query(filters)
+
+        if query:
+            self.filter_query.must.append(
+                ES_Q("query_string", query=query, default_operator="OR", fields=self.query_fields)
+            )
+
+        # Ensure that only non-zero values are taken into consideration
+        non_zero_queries = []
+        for field in self.sum_column_mapping.values():
+            non_zero_queries.append(ES_Q("range", **{field: {"gt": 0}}))
+            non_zero_queries.append(ES_Q("range", **{field: {"lt": 0}}))
+        self.filter_query.must.append(ES_Q("bool", should=non_zero_queries, minimum_should_match=1))
+
+        self.bucket_count = get_number_of_unique_terms_for_awards(self.filter_query, f"{self.agg_key}.hash")
+
+        results = self.query_elasticsearch()
+
+        return Response(
+            {
+                "results": results[: self.pagination.limit],
+                "page_metadata": get_pagination_metadata(
+                    self.bucket_count, self.pagination.limit, self.pagination.page
+                ),
+            }
+        )
+
+    @abstractmethod
+    def build_elasticsearch_result(self, response: dict) -> List[dict]:
+        pass
+
+    def build_elasticsearch_search_with_aggregations(self) -> Optional[AwardSearch]:
+        """
+        Using the provided ES_Q object creates an AwardSearch object with the necessary applied aggregations.
+        """
+        # Create the initial search using filters
+        search = AwardSearch().filter(self.filter_query)
+
+        pagination_values = {
+            "from": (self.pagination.page - 1) * self.pagination.limit,
+            "size": self.pagination.limit + 1,
+        }
+        sort_values = {"order": {self.sort_column_mapping[self.pagination.sort_key]: self.pagination.sort_order}}
+
+        if self.agg_key == settings.ES_ROUTING_FIELD:
+            size = self.pagination.upper_limit
+            shard_size = size
+            group_by_agg_key_values = {**sort_values}
+            bucket_sort_values = {**pagination_values}
+
+        else:
+            if self.bucket_count == 0:
+                return None
+            else:
+                size = self.bucket_count
+                shard_size = self.bucket_count + 100
+                group_by_agg_key_values = {}
+                bucket_sort_values = {**pagination_values, **sort_values}
+
+        if shard_size > 10000:
+            raise ElasticsearchConnectionException(
```

Comment:
Our exceptions aren't extensive, but this may not be the best to throw back. This would appear to be a suggestion that the client limit their query, and so should be a 4xx response, whereas this one gives a 500 implying something is wrong with our cluster. 

Maybe subclass? Or override the status code and details if possible.

Reply:
Plan to add a new exception for HTTP 403:

>The request was a legal request, but the server is refusing to respond to it. Unlike a 401 Unauthorized response, authenticating will make no difference.